### ISSUE TITLE
C++ make NamedAnyModule(name, any) public

### ISSUE THREAD

**Issue by meganset** [pytorch/pytorch#36720]
similar to issue #33192

if we start with AnyModule a, there is no good way to add a name, 
to create a NamedAnyModule
i.e. in  _nn/modules/container/named_any.h_

```
private:
  /// Creates a `NamedAnyModule` from a type-erased `AnyModule`.
  NamedAnyModule(std::string name, AnyModule any_module)
    : name_(std::move(name)), module_(std::move(any_module)) {}
```
I can create the pr to move this method to public if it makes sense.
(similar to pr #34208)

cc @yf225

**Comment by mruberry** [comment]
Hey @meganset, thanks for the suggestion! Can you add some context to this suggestion: when do you think users of PyTorch will want to do this, and what, if any, workaround(s) exist today for them to get that behavior? 

If there's a cool reason to have it and you're interested in submitting a PR for it that sounds good to me, although I'm not our nn expert.

**Comment by meganset** [comment]
Thanks for looking at this.

This change is similar to the one made in February with @yf225 (PR #34208)
where we made private methods in _Sequential_ using _AnyModule_’s public. 
in that case, if we had an _AnyModule_, there was no good way to add it to a _Sequential_ without recasting back to its original type.

Similarly, if we have an _AnyModule_, there's no way currently to make it into a _NamedAnyModule_ without recasting back to its original type (then the _NamedAnyModule_ constructors can recast it back to an _AnyModule_ & a name)
The public constructors call the private one that works with _AnyModule_ directly.

I submitted PR #36869 which moves the private constructor to the public section in _nn/modules/container/named_any.h_
